lib/chai/core/assertions.js

```diff
@@ -1268,120 +1268,114 @@ module.exports = function (chai, _) {
    */
 
   function assertThrows (constructor, errMsg, msg) {
-    if (msg) flag(this, 'message', msg);
-    var obj = flag(this, 'object');
-    new Assertion(obj, msg).is.a('function');
 
-    var thrown = false
-      , desiredError = null
-      , name = null
-      , thrownError = null;
-
-    if (arguments.length === 0) {
-      errMsg = null;
-      constructor = null;
-    } else if (constructor && (constructor instanceof RegExp || 'string' === typeof constructor)) {
-      errMsg = constructor;
-      constructor = null;
-    } else if (constructor && constructor instanceof Error) {
-      desiredError = constructor;
-      constructor = null;
-      errMsg = null;
-    } else if (typeof constructor === 'function') {
-      name = constructor.prototype.name;
-      if (!name || (name === 'Error' && constructor !== Error)) {
-        name = constructor.name || (new constructor()).name;
+    function stringify (expected) {
+      if (expected.e) {
+        return expected.e.message ? expected.e.toString() : expected.e.name;
+      }
+      if (expected.type) {
+        var constructor = expected.type;
+        return new constructor().name;
       }
-    } else {
-      constructor = null;
     }
 
-    try {
-      obj();
-    } catch (err) {
-      // first, check desired error
-      if (desiredError) {
-        this.assert(
-            err === desiredError
-          , 'expected #{this} to throw #{exp} but #{act} was thrown'
-          , 'expected #{this} to not throw #{exp}'
-          , (desiredError instanceof Error ? desiredError.toString() : desiredError)
-          , (err instanceof Error ? err.toString() : err)
-        );
-
-        flag(this, 'object', err);
-        return this;
+    function similarException (expected, thrown) {
+      var exception = expected.e;
+      if (exception && exception.constructor === thrown.constructor) {
+        return exception.name === thrown.name;
       }
+      return !expected.type || thrown instanceof expected.type;
+    }
 
-      // next, check constructor
-      if (constructor) {
-        this.assert(
-            err instanceof constructor
-          , 'expected #{this} to throw #{exp} but #{act} was thrown'
-          , 'expected #{this} to not throw #{exp} but #{act} was thrown'
-          , name
-          , (err instanceof Error ? err.toString() : err)
-        );
-
-        if (!errMsg) {
-          flag(this, 'object', err);
-          return this;
-        }
+    function similarMessage (expected, thrown) {
+      var isThrownedException = typeof thrown === 'object' && thrown.hasOwnProperty('message');
+      var message = isThrownedException ? thrown.message : '' + thrown;
```

Comment:
I'd prefer an explicit `String()` conversion here, as if a class has `valueOf()` it will be called over toString(), e.g.:

```js
'' + { valueOf() { return 1; }, toString() { return 'hello'; } }
```